### PR TITLE
feat: add SearchIOAnalytics as a simpler tracking solution

### PR DESCRIPTION
**Pull request by JasonBerry**
After discussion with @benhinchley around the new/old API I opted to keep the tracking module entirely self-contained and not rely on the Client to handle communication with the API.

### CODE REVIEW

File: src/tracking.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  account: string;
+  collection: string;
+  endpoint: string;
+  events: Record<Value, EventState[]>;
+  queryId?: string;
+
+  constructor(
+    account: string,
+    collection: string,
+    endpoint: string = "https://api.search.io"
```
Comment (wwalser) [inline src/tracking.ts:28]: Wondering if it's possible to just take a `Client` as the single operand here.

File: src/index.ts [hunk 2]
```diff
@@ -1240,13 +1241,15 @@ export type TokenState = {
...
 // Just here to handle SSR execution (docs)
-const setItem = (key: string, value: string) => {
+export const setItem = (key: string, value: string) => {
   if (isSSR()) {
     return;
   }
-  return localStorage.setItem(key, value);
+  try {
+    localStorage.setItem(key, value);
+  } catch (_) {}
```
Comment (wwalser) [inline src/index.ts:1250]: Let's check if console log (or console error) exists and if so, log something meaningful to it. Unlikely, but it's possible that a customer discovers this erroring out and reports it to us.

File: src/tracking.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  queryId: string;
+  type: string;
+  timestamp: number;
+  submitted: boolean;
+  metadata?: Metadata;
+}
+type Value = string | number;
+type Identifier = "result_id" | "redirect_id" | "promotion_id";
+
+export class SearchIOAnalytics {
```
Comment (wwalser) [inline src/tracking.ts:18]: We auto-generate documentation based on docblocks. You can do that later if you wish.

File: src/tracking.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  updateQueryID(queryId: string) {
+    this.queryId = queryId;
+  }
+
+  track(type: string, value: Value, metadata?: Metadata) {
+    let queryId = this.queryId;
+    if (!FUNNEL_ENTRY_TYPES.includes(type)) {
+      const lastEvent = this.getEvents(value).reverse().pop();
```
Comment (wwalser) [inline src/tracking.ts:198]: I think it'd be clearer to just lookup by index rather than calling two mutation functions.

**Review by wwalser — APPROVED** [review]
A few suggestions, but open handed on all of them.

File: src/tracking.ts (hunk 1 above)
Comment (JasonBerry) [reply src/tracking.ts:28]: I don't think it's worth introducing a dependency on `Client` purely for the `account` and `collection` values.

File: src/index.ts (hunk 2 above)
Comment (JasonBerry) [reply src/index.ts:1250]: Good idea. 👍

File: src/tracking.ts (hunk 3 above)
Comment (JasonBerry) [reply src/tracking.ts:18]: I'll add that now.

File: src/tracking.ts (hunk 4 above)
Comment (JasonBerry) [reply src/tracking.ts:198]: I uncovered a bug in this code due to your comment, the `reverse` mutation was incorrect.

**Pull request merged**